From ba2182196ec25c9f0bbbe92d63c7f54252288c8c Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Unload plugins after everything else
Make Durability enchant scale 0-80% reduction on Armor damage (up to level 5)
Zombies break down doors on normal
Update inventory on block place/interact cancellation
Don't update server.properties
Remove spigot update check
Supress empty itemstack log messages
---
 src/main/java/net/minecraft/server/EntityItem.java               | 2 +-
 src/main/java/net/minecraft/server/EntityOcelot.java             | 2 +-
 src/main/java/net/minecraft/server/ItemStack.java                | 9 +++++++++
 src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java  | 2 +-
 src/main/java/net/minecraft/server/PropertyManager.java          | 2 +-
 src/main/java/org/bukkit/craftbukkit/CraftServer.java            | 3 ++-
 .../java/org/bukkit/craftbukkit/event/CraftEventFactory.java     | 2 ++
 7 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 1f8b1c7..bd454c0 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -386,7 +386,7 @@ public class EntityItem extends Entity {
 
         if (itemstack == null) {
             if (this.world != null) {
-                EntityItem.b.error("Item entity " + this.getId() + " has no item?!");
+                //EntityItem.b.error("Item entity " + this.getId() + " has no item?!"); // EMC
             }
 
             return new ItemStack(Blocks.STONE);
diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index edcb307..c054012 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -6,7 +6,7 @@ public class EntityOcelot extends EntityTameableAnimal {
 
     private PathfinderGoalAvoidTarget<EntityHuman> bo;
     private PathfinderGoalTempt bp;
-    public boolean spawnBonus = true; // Spigot
+    public boolean spawnBonus = false; // Spigot // EMC
 
     public EntityOcelot(World world) {
         super(world);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 5496805..46be9b2 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -357,7 +357,16 @@ public final class ItemStack {
                 int j = EnchantmentManager.getEnchantmentLevel(Enchantment.DURABILITY.id, this);
                 int k = 0;
 
+                j = Math.min(5, j); // EMC
                 for (int l = 0; j > 0 && l < i; ++l) {
+                    // EMC start - make Unbreaking scale to level for armor. 5 = 80% chance of no loss.
+                    if (getItem() instanceof ItemArmor) {
+                        if (random.nextFloat() >= 1 - (.8 * (j / 5))) {
+                            k++;
+                        }
+                        continue;
+                    }
+                    // EMC end
                     if (EnchantmentDurability.a(this, j, random)) {
                         ++k;
                     }
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
index fdc77cb..4948867 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
@@ -62,7 +62,7 @@ public class PathfinderGoalBreakDoor extends PathfinderGoalDoorInteract {
             this.h = i;
         }
 
-        if (this.g == 240 && this.a.world.getDifficulty() == EnumDifficulty.HARD) {
+        if (this.g == 240 && this.a.world.getDifficulty().ordinal() >= EnumDifficulty.NORMAL.ordinal()) { // EMC
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.a, this.b.getX(), this.b.getY(), this.b.getZ()).isCancelled()) {
                 this.c();
diff --git a/src/main/java/net/minecraft/server/PropertyManager.java b/src/main/java/net/minecraft/server/PropertyManager.java
index 0ee7fe8..0c35e4f 100644
--- a/src/main/java/net/minecraft/server/PropertyManager.java
+++ b/src/main/java/net/minecraft/server/PropertyManager.java
@@ -72,7 +72,7 @@ public class PropertyManager {
 
         try {
             // CraftBukkit start - Don't attempt writing to file if it's read only
-            if (this.file.exists() && !this.file.canWrite()) {
+            if (this.file.exists()) { // EMC - Never overwrite the file
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f57c785..a95a4f9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -308,6 +308,7 @@ public final class CraftServer implements Server {
         if (type == PluginLoadOrder.STARTUP) {
             helpMap.clear();
             helpMap.initializeGeneralTopics();
+            loadCustomPermissions(); // EMC - Plugins should have access to this
         }
 
         Plugin[] plugins = pluginManager.getPlugins();
@@ -325,7 +326,7 @@ public final class CraftServer implements Server {
             setVanillaCommands(false);
             // Spigot end
             commandMap.registerServerAliases();
-            loadCustomPermissions();
+            // loadCustomPermissions(); // EMC - Moving up
             DefaultPermissions.registerCorePermissions();
             CraftDefaultPermissions.registerCorePermissions();
             helpMap.initializeCommands();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index d1bd480..b3d3a9c 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -125,6 +125,7 @@ public class CraftEventFactory {
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, player.getItemInHand(), player, canBuild);
         craftServer.getPluginManager().callEvent(event);
 
+        if (event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
@@ -226,6 +227,7 @@ public class CraftEventFactory {
         }
         craftServer.getPluginManager().callEvent(event);
 
+        if (action == Action.RIGHT_CLICK_BLOCK && event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
-- 
1.9.1

